From 053058d9027af744839e969d84cd65a56cb36161 Mon Sep 17 00:00:00 2001
From: "Christopher A. Williamson" <contact@chrisaw.io>
Date: Sun, 22 Feb 2026 16:32:11 +0000
Subject: [PATCH 2/3] Add Unihertz/Agold hardware control tool (agold-cmd)

Update agold-cmd to support ioctl and cmd subcommands for controlling
Unihertz 8849 device hardware (projector, lights, sensors). The cmd()
AIDL method takes an
IAgoldDaemonCallback, so implement BnAgoldDaemonCallback with an
onResult() handler that prints the daemon's response.

Requires the vendor_interfaces AIDL fix that corrects cmd()'s callback
parameter type from IAgoldDaemon to IAgoldDaemonCallback.

Include agold-cmd in PRODUCT_PACKAGES so it ships in the GSI image.
---
 base.mk            |  3 ++-
 cmds/agold-cmd.cpp | 32 ++++++++++++++++++++------------
 2 files changed, 22 insertions(+), 13 deletions(-)

diff --git a/base.mk b/base.mk
index a29e402..9314eba 100644
--- a/base.mk
+++ b/base.mk
@@ -147,7 +147,8 @@ PRODUCT_PACKAGES += \
 PRODUCT_PACKAGES += \
     lightsctl \
     lightsctl-aidl \
-    uevent
+    uevent \
+    agold-cmd
 
 PRODUCT_COPY_FILES += \
     device/phh/treble/files/adbd.rc:system/etc/init/adbd.rc
diff --git a/cmds/agold-cmd.cpp b/cmds/agold-cmd.cpp
index 9943557..cd19275 100644
--- a/cmds/agold-cmd.cpp
+++ b/cmds/agold-cmd.cpp
@@ -1,22 +1,30 @@
 #include <iostream>
 #include <vendor/mediatek/hardware/agolddaemon/IAgoldDaemon.h>
+#include <vendor/mediatek/hardware/agolddaemon/BnAgoldDaemonCallback.h>
 #include <binder/IServiceManager.h>
+#include <utils/String8.h>
 
 using ::vendor::mediatek::hardware::agolddaemon::IAgoldDaemon;
+using ::vendor::mediatek::hardware::agolddaemon::BnAgoldDaemonCallback;
 using ::android::sp;
 
-int main(int argc, char **argv) {
-	auto svc = android::checkVintfService<IAgoldDaemon>();
-    android::String16 a;
-    svc->getNotGsi(&a);
-    std::cout << "notgsi is " << a << std::endl;
-    if (argc >= 2) {
-        android::String16 b(argv[1]);
-        svc->setNotGsi(b);
-    }
-    svc->getNotGsi(&a);
-    std::cout << "notgsi is " << a << std::endl;
+class AgoldDaemonCallback : public BnAgoldDaemonCallback {
+    public:
+        ::android::binder::Status onResult(const ::android::String16& result) override {
+            std::cout << "Result: " << ::android::String8(result).c_str() << std::endl;
+            return ::android::binder::Status::ok();
+        }
+};
 
-    svc->setCameraClientPackageName(android::String16("com.mediatek.camera"));
+int main(int argc [[maybe_unused]], char **argv) {
+	auto svc = android::checkVintfService<IAgoldDaemon>();
+    int32_t ret = 0;
 
+    if (std::string(argv[1]) == "ioctl") {
+        svc->SendMessageToIoctl(atoi(argv[2]), atoi(argv[3]), atoi(argv[4]), atoi(argv[5]), &ret);
+    } else {
+        sp<AgoldDaemonCallback> cb = new AgoldDaemonCallback();
+        svc->cmd(android::String16(argv[2]), cb);
+    }
+    printf("Returned %d\n", ret);
 }
-- 
2.52.0

